server/src-lib/Hasura/RQL/DDL/Remote/Validate.hs

```diff
@@ -0,0 +1,188 @@
+fieldNameToVariable :: FieldName -> G.Variable
+fieldNameToVariable = G.Variable . G.Name . getFieldNameTxt
+
+-- | Lookup the field in the schema.
+lookupField ::
+     G.Name
+  -> VT.ObjTyInfo
+  -> Either (NonEmpty ValidationError) VT.ObjFldInfo
+lookupField name objFldInfo = viaObject objFldInfo
+  where
+    viaObject =
+      maybe (Left (pure (CouldntFindRemoteField name objFldInfo))) pure .
+      HM.lookup name .
+      VT._otiFields
+
+-- | Validate remote input arguments against the remote schema.
+validateRemoteArguments ::
+     HashMap G.Name InpValInfo
+  -> HashMap G.Name G.Value
+  -> HashMap G.Variable FieldInfo
+  -> Validation (NonEmpty ValidationError) ()
+validateRemoteArguments expectedArguments providedArguments permittedVariables = do
+  traverse validateProvided (HM.toList providedArguments)
+  traverse validateExpected (HM.toList expectedArguments)
+  pure ()
+  where
+    validateProvided (providedKey, providedValue) =
+      case HM.lookup providedKey expectedArguments of
+        Nothing -> Failure (pure (NoSuchArgumentForRemote providedKey))
+        Just expectedType ->
+          validateType permittedVariables providedValue (_iviType expectedType)
+    validateExpected (expectedKey, expectedInpValInfo) =
+      if G.isNullable (_iviType expectedInpValInfo)
+        then pure ()
+        else case _iviDefVal expectedInpValInfo of
+               Just {} -> pure ()
+               Nothing ->
+                 case HM.lookup expectedKey providedArguments of
+                   Nothing ->
+                     Failure (pure (MissingRequiredArgument expectedKey))
+                   Just {} -> pure ()
+
+-- | Validate a value against a type.
+validateType ::
+     HashMap G.Variable FieldInfo
+  -> G.Value
+  -> G.GType
+  -> Validation (NonEmpty ValidationError) ()
+validateType permittedVariables value (gtypeToEither -> expectedType) =
+  bindValidation
+    (valueType permittedVariables value)
+    (\actualType ->
+       when (actualType /= expectedType) (Failure (pure InvalidType)))
+
+-- | Produce the type of a value.
+valueType ::
+     HashMap G.Variable FieldInfo
+  -> G.Value
+  -> Validation (NonEmpty ValidationError) (Either G.NamedType G.ListType)
+valueType permittedVariables =
+  \case
+    G.VVariable variable ->
+      case HM.lookup variable permittedVariables of
+        Nothing -> Failure (pure (InvalidVariable variable permittedVariables))
+        Just fieldInfo -> fmap Left (fieldInfoToNamedType fieldInfo)
+    G.VInt {} -> pure (Left (mkScalarTy PGInteger))
+    G.VFloat {} -> pure (Left (mkScalarTy PGFloat))
+    G.VBoolean {} -> pure (Left (mkScalarTy PGBoolean))
+    G.VNull -> Failure (pure NullNotAllowedHere)
+    G.VString {} -> pure (Left (mkScalarTy PGText))
+    (G.VEnum _) -> error "TODO: G.VEnum: What do we do here?"
+    (G.VList _) -> error "TODO: G.VList: What do we do here?"
+    (G.VObject _) -> error "TODO: G.VObject: What do we do here?"
```

Comment:
We will need to support object to support remote arguments of this form:

```
type: create_remote_relationship
args:
  name: message
  table: profiles
  hasura_fields:
    - id
  remote_schema: my-remote-schema
  remote_field: messages
  remote_arguments:
    where:
      id:
        _eq: "$id"
```

Reply:
This is going to be a bit complex, let's throw a `NotSupported` validation error for now. And get back to this later.

Reply:
I think that if I extend the python fake graphql service to expect argument format like that, then it'll be easier to test for objects. Will we support lists? I don't see a reason not to, but you might? Will we support enums? 

For later reference, can you point to an example of how I might do that on the python side? I'm not sure what to do in [the python here](https://github.com/hasura/graphql-engine/blob/500089308832d15bf2f29583b2f227226c0f2526/server/tests-py/graphql_server.py#L611). I'm confident I could figure out the Haskell side.

Reply:
True. We might need to support everything. For e.g. what if the join query is something like:

```
messages(include: [$id]) {
   id
   name
}
```


I can extend the python server to have more complicated types.

Reply:
Added the validation error now. I've copied over our conversation to the issue tracker: https://github.com/tirumaraiselvan/graphql-engine/issues/9